libyul/optimiser/FunctionCallFinder.cpp

```diff
@@ -16,23 +16,46 @@
 */
 
 #include <libyul/optimiser/FunctionCallFinder.h>
+
+#include <libyul/optimiser/ASTWalker.h>
 #include <libyul/AST.h>
 
 using namespace solidity;
 using namespace solidity::yul;
 
-std::vector<FunctionCall*> FunctionCallFinder::run(Block& _block, YulString _functionName)
+namespace
+{
+template<typename Base, typename ResultType>
+class MaybeConstFunctionCallFinder: Base
 {
-	FunctionCallFinder functionCallFinder(_functionName);
-	functionCallFinder(_block);
-	return functionCallFinder.m_calls;
+public:
+	using MaybeConstBlock = std::conditional_t<std::is_const_v<ResultType>, Block const, Block>;
+	static std::vector<ResultType*> run(MaybeConstBlock& _block, YulString const _functionName)
+	{
+		MaybeConstFunctionCallFinder functionCallFinder(_functionName);
+		functionCallFinder(_block);
+		return functionCallFinder.m_calls;
+	}
+private:
+	explicit MaybeConstFunctionCallFinder(YulString const _functionName): m_functionName(_functionName), m_calls() {}
+	using Base::operator();
+	void operator()(ResultType& _functionCall) override
+	{
+		Base::operator()(_functionCall);
+		if (_functionCall.functionName.name == m_functionName)
+			m_calls.emplace_back(&_functionCall);
+	}
+	YulString m_functionName;
+	std::vector<ResultType*> m_calls;
+};
 }
 
-FunctionCallFinder::FunctionCallFinder(YulString _functionName): m_functionName(_functionName) {}
+std::vector<FunctionCall*> solidity::yul::findFunctionCalls(Block& _block, YulString _functionName)
+{
+	return MaybeConstFunctionCallFinder<ASTModifier, FunctionCall>::run(_block, _functionName);
```

Comment:
Are there any other things that we may need to collect and preserve constness?

Reply:
At a quick glance I see that we do have some other AST finders: `MSizeFinder`, `InlinableExpressionFunctionFinder`, `LeaveFinder`, `TerminationFinder`. Haven't checked which of them would fit this pattern though.